linalg/sparsemat.cpp

```diff
@@ -2145,31 +2145,49 @@ void SparseMatrix::DiagScale(const Vector &b, Vector &x, double sc) const
 {
    MFEM_VERIFY(Finalized(), "Matrix must be finalized.");
 
+   const int nnz = J.Capacity();
+
+   const bool use_dev = b.UseDevice() || x.UseDevice();
+
+   auto bp = b.Read(use_dev);
+   auto xp = x.Write(use_dev);
+
+   auto Ap = Read(A, nnz);
+   auto Ip = Read(I, height+1);
+   auto Jp = Read(J, nnz);
+
    bool scale = (sc != 1.0);
-   for (int i = 0, j = 0; i < height; i++)
+   MFEM_FORALL(i, height,
    {
-      int end = I[i+1];
-      for ( ; true; j++)
+      int end = Ip[i+1];
+      for (int j = Ip[i]; true; j++)
       {
-         MFEM_VERIFY(j != end, "Couldn't find diagonal in row. i = " << i
-                     << ", j = " << j
-                     << ", I[i+1] = " << end );
-         if (J[j] == i)
+         if (j == end)
          {
-            MFEM_VERIFY(std::abs(A[j]) > 0.0, "Diagonal " << j << " must be nonzero");
+            printf("Couldn't find diagonal in row. i = %d, j = %d, I[i+1] = %d\n",
+            i, j, end);
```

Comment:
I just now see this: using `printf` violates one of the xSDK requirements (or was this just recommendation?) -- a user of the library should be able to capture or turn off any output from the library. This is why we introduced `mfem::out` and `mfem::err`.

GPUs are somewhat different, so for now it may be okay to just use `printf` on GPU but we should be able to use `mfem::err` on CPU. One way is to change the `MFEM_ABORT_KERNEL` macro to use `printf`-like input. On GPU, the macro can simply pass the arguments to `printf` and then abort. On CPU, the arguments can be passed to `sprintf` first and then the resulting string can be passed to `MFEM_ABORT`.

I think it is probably best to take the time now to address this issue, instead of postponing it.

Reply:
@v-dobrev Thanks for the suggestions, which are incorporated in the latest commit. I like the idea of using `sprintf` to pass a string to `MFEM_ABORT_KERNEL` if necessary, but in this case I just simplified the error message by not including the data (`i,j`, etc). Such detailed data can be found with a debugger.